crmprtd/insert.py

```diff
@@ -142,85 +178,133 @@ def bisect_insert_strategy(sesh, obs):
             sesh.rollback()
 
-            a, b = split(obs)
+            a, b = split(observations)
             dbm_a = bisect_insert_strategy(sesh, a)
             dbm_b = bisect_insert_strategy(sesh, b)
             return dbm_a + dbm_b
         else:
             log.info(
-                f"Successfully inserted observations: {len(obs)}",
-                extra={"num_obs": len(obs)},
+                f"Successfully inserted observations: {len(observations)}",
+                extra={"num_obs": len(observations)},
             )
-            db_metrics = DBMetrics(len(obs), 0, 0)
+            db_metrics = DBMetrics(len(observations), 0, 0)
         sesh.commit()
         return db_metrics
 
 
-def insert(sesh, observations, sample_size):
+def bulk_insert_strategy(sesh, observations):
+    """
+    This method performs a bulk insert of observations using the PostgreSQL dialect
+    INSERT ... ON CONFLICT DO IGNORE clause that allows bulk inserts with duplicates
+    to proceed without raising an exception (dups are ignored; new rows are inserted).
+
+    NOTE: This only works for PostgreSQL databases. Other dialects may support a
+    similar operation, but this code does not handle those cases.
+
+    :param sesh: SQLAlchemy database session
+    :param observations: List of observations to insert
+    :return: DMMetrics describing result of insertion
+    """
+    num_to_insert = len(observations)
+
+    # If you try to insert no rows with pg_insert, there's an unexpected insertion.
+    # Avoid that.
+    if num_to_insert == 0:
+        return DBMetrics(0, 0, 0)
+
+    try:
+        with sesh.begin_nested():
+            result = sesh.execute(
+                pg_insert(Obs)
+                .values(observations)
+                .on_conflict_do_nothing()
+                .returning(Obs.id)
+            ).fetchall()
+    except DBAPIError as e:
+        # Something really unanticipated happened. Duplicate rows do not trigger an
+        # exception.
+        log.warning(f"Unexpected error during bulk insertion: {e}")
+        return DBMetrics(0, 0, num_to_insert)
+    else:
+        num_inserted = len(result)
+        log.info(
+            f"Successfully inserted observations: {num_inserted}",
+            extra={"num_obs": num_inserted},
+        )
+    sesh.commit()
+    return DBMetrics(num_inserted, num_to_insert - num_inserted, 0)
+
+
+def insert(
+    sesh,
+    observations,
+    strategy=InsertStrategy.BULK,
+    bulk_chunk_size=1000,
+    sample_size=50,
+):
     """
     Insert a collection of observations.
 
     :param sesh: SQLAlchemy database session
-    :param observations: Obs objects to insert
-    :param sample_size: Size of sample of observations to use to test if all duplicates.
+    :param observations: For BULK insert strategy, list of dicts containing Obs rows
+        to insert. For OTHER insert strategy, list of Obs objects to insert.
+    :param strategy: Insert strategy. BULK is currently fastest.
```

Comment:
Oops, that should have been BISECTION. The list of options is defined by the enum InsertStrategy. I don't think it's a good idea to repeat it here, since if it changes this derived string would have to change. I'll put in a note "see InsertStrategy".